packages/graph/hash_graph/postgres_migrations/V1__initial.sql

```diff
@@ -0,0 +1,270 @@
+  "version" BIGINT NOT NULL,
+  "version_id" UUID REFERENCES "version_ids"
+);
+
+COMMENT ON TABLE "type_ids" IS $pga$ This table is a boundary to define the actual identification scheme for our kinds of types. Assume that we use the UUIDs on the types to look up more specific ID details. $pga$;
+
+ALTER TABLE "type_ids"
+  ADD CONSTRAINT "type_ids_primary_key" PRIMARY KEY ("base_uri", "version");
+
+CREATE TABLE IF NOT EXISTS "data_types" (
+  "version_id" UUID PRIMARY KEY REFERENCES "version_ids",
+  "schema" JSONB NOT NULL,
+  "owned_by_id" UUID NOT NULL REFERENCES "accounts",
+  "updated_by_id" UUID NOT NULL REFERENCES "accounts"
+);
+
+CREATE TABLE IF NOT EXISTS "property_types" (
+  "version_id" UUID PRIMARY KEY REFERENCES "version_ids",
+  "schema" JSONB NOT NULL,
+  "owned_by_id" UUID NOT NULL REFERENCES "accounts",
+  "updated_by_id" UUID NOT NULL REFERENCES "accounts"
+);
+
+CREATE TABLE IF NOT EXISTS "entity_types" (
+  "version_id" UUID PRIMARY KEY REFERENCES "version_ids",
+  "schema" JSONB NOT NULL,
+  "owned_by_id" UUID NOT NULL REFERENCES "accounts",
+  "updated_by_id" UUID NOT NULL REFERENCES "accounts"
+);
+
+CREATE TABLE IF NOT EXISTS "property_type_property_type_references" (
+  "source_property_type_version_id" UUID NOT NULL REFERENCES "property_types",
+  "target_property_type_version_id" UUID NOT NULL REFERENCES "property_types"
+);
+
+CREATE TABLE IF NOT EXISTS "property_type_data_type_references" (
+  "source_property_type_version_id" UUID NOT NULL REFERENCES "property_types",
+  "target_data_type_version_id" UUID NOT NULL REFERENCES "data_types"
+);
+
+CREATE TABLE IF NOT EXISTS "entity_type_property_type_references" (
+  "source_entity_type_version_id" UUID NOT NULL REFERENCES "entity_types",
+  "target_property_type_version_id" UUID NOT NULL REFERENCES "property_types"
+);
+
+CREATE TABLE IF NOT EXISTS "entity_type_entity_type_references" (
+  "source_entity_type_version_id" UUID NOT NULL REFERENCES "entity_types",
+  "target_entity_type_version_id" UUID NOT NULL REFERENCES "entity_types"
+);
+
+CREATE TABLE IF NOT EXISTS "entity_ids" (
+  "owned_by_id" UUID NOT NULL REFERENCES "accounts",
+  "entity_uuid" UUID NOT NULL,
+  "left_owned_by_id" UUID,
+  "left_entity_uuid" UUID,
+  "right_owned_by_id" UUID,
+  "right_entity_uuid" UUID
+);
+
+ALTER TABLE "entity_ids"
+  ADD CONSTRAINT "entity_ids_primary_key" PRIMARY KEY ("owned_by_id", "entity_uuid"),
+  ADD CONSTRAINT "entity_ids_left_reference" FOREIGN KEY ("left_owned_by_id", "left_entity_uuid") REFERENCES "entity_ids",
+  ADD CONSTRAINT "entity_ids_right_reference" FOREIGN KEY ("right_owned_by_id", "right_entity_uuid") REFERENCES "entity_ids",
+  ADD CONSTRAINT "entity_ids_relation_constraint" CHECK (left_entity_uuid IS NULL AND right_entity_uuid IS NULL AND left_owned_by_id IS NULL AND right_owned_by_id IS NULL     OR        left_entity_uuid IS NOT NULL AND right_entity_uuid IS NOT NULL AND left_owned_by_id IS NOT NULL AND right_owned_by_id IS NOT NULL);
```

Comment:
It may be worth splitting this across lines
```suggestion
  ADD CONSTRAINT "entity_ids_relation_constraint" CHECK (left_entity_uuid IS NULL AND right_entity_uuid IS NULL AND left_owned_by_id IS NULL AND right_owned_by_id IS NULL OR left_entity_uuid IS NOT NULL AND right_entity_uuid IS NOT NULL AND left_owned_by_id IS NOT NULL AND right_owned_by_id IS NOT NULL);
```

Reply:
3044e5dbf1116c098d3f00b7325dd0623d5d488d